src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -2572,6 +2631,80 @@ address StubGenerator::generate_base64_decodeBlock() {
     __ BIND(L_bruteForce);
   }   // End of if(avx512_vbmi)
 
+  if (VM_Version::supports_avx2()) {
+    Label L_tailProc, L_topLoop, L_enterLoop;
+    ////////////////////////////////////////////
+    //
+    //  **** Make sure inputs match ****
+    // size_t fast_avx2_base64_decode(char *out, const char *src, size_t srclen)
+    //
+    // Handle isURL / MIME?!?!  r12 is used for length calculation (from out)
+    //
+    // rbx is out, r12 is saved out, rdx is size, rsi is src
+    //
+    ////////////////////////////////////////////
+
+    // *************** NEEDS TO BE FIXED ************
+    __ cmpl(isURL, 0);
+    __ jcc(Assembler::notZero, L_tailProc);
+
+    __ cmpl(length, 44);
+    __ jcc(Assembler::belowEqual, L_tailProc);
+
+    __ cmpl(isMIME, 0);
+    __ jcc(Assembler::notEqual, L_tailProc);
+
+    // Set up constants
+    __ lea(r13, ExternalAddress(StubRoutines::x86::base64_AVX2_tables_addr()));
+    __ vpbroadcastq(xmm4, Address(r13, 0), Assembler::AVX_256bit);
+    __ vmovdqu(xmm11, Address(r13, 0x28));
+
+    // The 0x2f2f..2f vector is used both as a table and a mask.  For now, just initialize
+    // using movq instead of a broadcast (slight performance increase).
+    // Will likely be changed back when URL is accelerated.
+    // __ vpbroadcastb(xmm10, Address(r13, 0), Assembler::AVX_256bit);
+    __ vmovdqu(xmm10, xmm4);
+    __ vmovdqu(xmm9, Address(r13, 0x48));
+    __ vmovdqu(xmm8, Address(r13, 0x68));
+    __ vmovdqu(xmm7, Address(r13, 0x88));
+    __ vmovdqu(xmm6, Address(r13, 0xa8));
+    __ vmovdqu(xmm13, Address(r13, 0xc8));
+    __ vmovdqu(xmm12, Address(r13, 0x08));
+    __ jmp(L_enterLoop);
```

Comment:
This got me curious (sorry!) and maybe there's something going on here that I'm not getting... But why have you split the loop apart and jump into the middle of it? It'd seem not doing this would be more straightforward, with no difference semantically and one less jump? (align32 should just translate to a narrow nop instruction, if anything)